lib/swm050/gpio.c

```diff
@@ -0,0 +1,173 @@
+*/
+void gpio_toggle(uint16_t gpios)
+{
+	uint32_t curr_status = GPIO_DATA & gpios;
+	GPIO_DATA = (GPIO_DATA & (~gpios)) | (~curr_status);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Set the direction of a Group of Pins
+
+Set the direction of one or more pins of GPIO.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	     If multiple pins are to be changed, use bitwise OR '|' to separate
+	     them.
+@param[in] dir Unsigned int8. Pin direction @ref gpio_dir
+*/
+void gpio_dir(uint16_t gpios, uint8_t dir)
+{
+	if (dir == GPIO_INPUT)
+		GPIO_DIR &= ~gpios;
+	else if (dir == GPIO_OUTPUT)
+		GPIO_DIR |= gpios;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Select the alternative function of a Group of Pins
+
+Select the alternative function of one or more pins of GPIO.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	     If multiple pins are to be changed, use bitwise OR '|' to separate
+	     them.
+@param[in] af_en Bool. Whether alternative function is selected
+*/
+void gpio_sel_af(uint16_t gpios, bool af_en)
+{
+	if (gpios & GPIO0)
+		GPIO_SEL = (GPIO_SEL & (~0x3)) | (af_en ? 0x1 : 0x0);
+	if (gpios & GPIO1)
+		GPIO_SEL = (GPIO_SEL & (~0xc)) | (af_en ? 0x4 : 0x0);
+	if (gpios & GPIO2)
+		GPIO_SEL = (GPIO_SEL & (~0x30)) | (af_en ? 0x10 : 0x0);
+	if (gpios & GPIO7)
+		GPIO_SEL = (GPIO_SEL & (~0xc000)) | (af_en ? 0x4000 : 0x0);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Enable the internal pull-up of a Group of Pins
+
+Enable or disable the internal pull-up of one or more pins of GPIO.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	     If multiple pins are to be changed, use bitwise OR '|' to separate
+	     them.
+@param[in] en Bool. Whether pull-up is enabled
+*/
+void gpio_pullup(uint16_t gpios, bool en)
+{
+	if (en)
+		GPIO_PULLUP |= gpios;
+	else
+		GPIO_PULLUP &= ~gpios;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Enable the input function of a Group of Pins
+
+Enable or disable the input function of one or more pins of GPIO. Disabling
+the input function of pins can decrease the power usage of the MCU.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	     If multiple pins are to be changed, use bitwise OR '|' to separate
+	     them.
+@param[in] en Bool. Whether input function is enabled
+*/
+void gpio_in_en(uint16_t gpios, bool en)
+{
+	if (en)
+		GPIO_INEN &= ~gpios;
+	else
+		GPIO_INEN |= gpios;
+}
+
+
+/*---------------------------------------------------------------------------*/
+/** @brief Enable the SWD debugging port
+
+Enable or disable the SWD debugging port. When SWD debugging port is enabled,
+GPIO and AF of the SWD pins will be both unavailable.
+
+@param[in] en Bool. Whether SWD is enabled
+*/
+void swd_en(bool en)
```

Comment:
I feel like this should probably be in a different file.  We try and keep the namespaces somewhat clear for each file too.  alternatively, prefix this with gpio_

Reply:
This is useful for enabling the GPIO under SWD pins.

Maybe it should be named something like gpio_switch_from_swd ?